storage/innobase/handler/i_s.cc

```diff
@@ -2660,100 +2657,62 @@ struct st_maria_plugin	i_s_innodb_ft_index_cache =
-
-	DBUG_EXECUTE_IF("fts_instrument_result_cache_limit",
-	        fts_result_cache_limit = 8192;
-	);
-
-	trx = trx_create();
-
-	trx->op_info = "fetching FTS index nodes";
-
-	pars_info_bind_function(info, "my_func", fetch.read_record, &fetch);
-	pars_info_bind_varchar_literal(info, "word", word->f_str, word->f_len);
-
-	FTS_INIT_INDEX_TABLE(&fts_table, fts_get_suffix(selected),
-			     FTS_INDEX_TABLE, index);
-	fts_get_table_name(&fts_table, table_name);
-	pars_info_bind_id(info, "table_name", table_name);
-
-	graph = fts_parse_sql(
-		&fts_table, info,
-		"DECLARE FUNCTION my_func;\n"
-		"DECLARE CURSOR c IS"
-		" SELECT word, doc_count, first_doc_id, last_doc_id,"
-		" ilist\n"
-		" FROM $table_name WHERE word >= :word;\n"
-		"BEGIN\n"
-		"\n"
-		"OPEN c;\n"
-		"WHILE 1 = 1 LOOP\n"
-		"  FETCH c INTO my_func();\n"
-		"  IF c % NOTFOUND THEN\n"
-		"    EXIT;\n"
-		"  END IF;\n"
-		"END LOOP;\n"
-		"CLOSE c;");
-
-	for (;;) {
-		error = fts_eval_sql(trx, graph);
-
-		if (UNIV_LIKELY(error == DB_SUCCESS)) {
-			fts_sql_commit(trx);
-
-			break;
-		} else {
-			fts_sql_rollback(trx);
-
-			if (error == DB_LOCK_WAIT_TIMEOUT) {
-				ib::warn() << "Lock wait timeout reading"
-					" FTS index. Retrying!";
-
-				trx->error_state = DB_SUCCESS;
-			} else {
-				ib::error() << "Error occurred while reading"
-					" FTS index: " << error;
-				break;
-			}
-		}
-	}
-
-	que_graph_free(graph);
-
-	trx->free();
-
-	if (fetch.total_memory >= fts_result_cache_limit) {
-		error = DB_FTS_EXCEED_RESULT_CACHE_LIMIT;
-	}
+dberr_t i_s_fts_index_table_fill_selected(
+  dict_index_t *index, ib_vector_t *words, ulint selected,
+  fts_string_t *word)
+{
+  dberr_t error= DB_SUCCESS;
+  ulint   total_memory= 0;
+  DBUG_EXECUTE_IF("fts_instrument_result_cache_limit",
+                  fts_result_cache_limit = 8192;);
+  trx_t* trx= trx_create();
+  trx->op_info= "fetching FTS index nodes";
+  for (;;)
+  {
+    FTSQueryExecutor executor(trx, index, index->table);
+    AuxRecordReader reader(words, &total_memory);
+    if (word->f_str == nullptr)
+      error= executor.read_aux_all((uint8_t) selected, reader);
+    else
+      error= executor.read_aux(
+        (uint8_t) selected,
+        reinterpret_cast<const char*>(word->f_str),
+        PAGE_CUR_GE, reader);
 
-	return(error);
+    if (UNIV_LIKELY(error == DB_SUCCESS ||
+                    error == DB_RECORD_NOT_FOUND))
+    {
+      fts_sql_commit(trx);
+      if (error == DB_RECORD_NOT_FOUND) error = DB_SUCCESS;
+      break;
+    }
+    else
+    {
+      fts_sql_rollback(trx);
+      if (error == DB_LOCK_WAIT_TIMEOUT)
+      {
+        ib::warn() << "Lock wait timeout reading FTS index. Retrying!";
+        trx->error_state = DB_SUCCESS;
+      }
+      else
+      {
+        ib::error() << "Error occurred while reading FTS index: " << error;
+        break;
+      }
+    }
+  }
+  trx->free();
+  if (total_memory >= fts_result_cache_limit)
+    error= DB_FTS_EXCEED_RESULT_CACHE_LIMIT;
```

Comment:
Shouldn’t this condition be checked inside the loop? Shouldn’t an `executor` member function return this error?